src/tactic/nth_rewrite/default.lean

```diff
@@ -0,0 +1,122 @@
+
+## Implementation details
+
+There are two alternative backends, provided by `.congr` and `.kabstract`.
+The kabstract backend is not currently available through mathlib.
+
+The kabstract backend is faster, but if there are multiple identical occurences of the
+same rewritable subexpression, all are rewritten simultaneously,
+and this isn't always what we want.
+(In particular, `rewrite_search` is much less capable on the `category_theory` library.)
+-/
+
+open tactic lean.parser interactive interactive.types expr
+
+namespace tactic
+
+/-- Returns the target of the goal when passed `none`,
+otherwise, return the type of `h` in `some h`. -/
+meta def target_or_hyp_type : option expr → tactic expr
+| none     := target
+| (some h) := infer_type h
+
+/-- Replace the target, or a hypothesis, depending on whether `none` or `some h` is given as the
+first argument. -/
+meta def replace_in_state : option expr → expr → expr → tactic unit
+| none     := tactic.replace_target
+| (some h) := λ e p, tactic.replace_hyp h e p >> skip
+
+open nth_rewrite nth_rewrite.congr nth_rewrite.tracked_rewrite
+open tactic.interactive
+
+/-- Preprocess a rewrite rule for use in `get_nth_rewrite`. -/
+private meta def unpack_rule (p : rw_rule) : tactic (expr × bool) :=
+do r ← to_expr p.rule tt ff,
+   return (r, p.symm)
+
+/-- Get the `n`th rewrite of rewrite rules `q` in expression `e`,
+or fail if there are not enough such rewrites. -/
+meta def get_nth_rewrite (n : ℕ) (q : rw_rules_t) (e : expr) : tactic tracked_rewrite :=
+do rewrites ← q.rules.mmap $ λ r, unpack_rule r >>= nth_rewrite e,
+   rewrites.join.nth n <|> fail "failed: not enough rewrites found"
+
+/-- Rewrite the `n`th occurence of the rewrite rules `q` of (optionally after zooming into) a
+hypothesis or target `h` which is an application of a relation. -/
+meta def get_nth_rewrite_with_zoom
+  (n : ℕ) (q : rw_rules_t) (path : list expr_lens.dir) (h : option expr) : tactic tracked_rewrite :=
+do e ← target_or_hyp_type h,
+   (ln, new_e) ← expr_lens.entire.zoom path e,
+   rw ← get_nth_rewrite n q new_e,
+   return ⟨ln.fill rw.exp, rw.proof >>= ln.congr, rw.addr.map $ λ l, path ++ l⟩
+
+/-- Rewrite the `n`th occurence of the rewrite rules `q` (optionally on a side)
+at all the locations `loc`. -/
+meta def nth_rewrite_core (path : list expr_lens.dir) (n : parse small_nat) (q : parse rw_rules)
+  (l : parse location) : tactic unit :=
+do let fn := λ h, get_nth_rewrite_with_zoom n q path h
+                    >>= λ rw, (rw.proof >>= replace_in_state h rw.exp),
+   match l with
+   | loc.wildcard := l.try_apply (fn ∘ some) (fn none)
+   | _            := l.apply     (fn ∘ some) (fn none)
+   end,
+   tactic.try (tactic.reflexivity reducible),
+   (returnopt q.end_pos >>= save_info <|> skip)
+
+namespace interactive
+
+open expr_lens
+
+/-- `nth_write n rules` performs only the `n`th possible rewrite using the `rules`.
```

Comment:
I think this could use some clarification. What's the "nth rewrite" if I give a list of rules? Is it the nth occurrence for each rule? Expanding this with a few examples would be nice, since it will become the tactic doc entry.

Reply:
I've tried to do this.